### ISSUE TITLE
rustbuild: x.py dist  should interleave tool builds 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#107450]
Right now doing a `x.py dist` roughly acts like this:
1) build tools deps // good parallelism
2) build tool // bad parallelism
3) thinlto tool // good parallelism
4) create dist artifact of the tool // bad parallelism 
5) proceed to next tool

steps 2 and 4 usually start initially with a number of N threads but these might be 8 or just two, and we fully wait until everything is done until processing with the next tool. 

I have an 8c/16t system and eyeballing `htop` while running `x.py dist` I would guess we could save perhaps 30%-50% wall time by `doing more stuff at the same at the same time`™ (saturating more cores).
However measurements were taken AFTER running a full `x.py build` so the heaving lifting/ building of rustc, cargo, rustfmt  and other tools.etc was already done.

**Comment by matthiaskrgr** [comment]
I checked if we could simply use more threads to compress the dist artifacts but it seems that throwing a higher thread limit at the problem doesn't really have any huge effect :(
https://github.com/rust-lang/rust-installer/blob/master/src/compression.rs#L42

**Comment by jyn514** [comment]
cc https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/async.20in.20bootstrap.3F, @the8472

this seems *nice*, but hard to do in a way that ensures correctness :/

**Comment by the8472** [comment]
Yep, can confirm that dist steps have some dips in core-utilization, especially the tarball creation.

> I checked if we could simply use more threads to compress the dist artifacts but it seems that throwing a higher thread limit at the problem doesn't really have any huge effect :(

Are you building both gz+xz artifacts? The gz branch seems single-threaded and be the bottleneck. 

> this seems nice, but hard to do in a way that ensures correctness :/

For dist specifically we might be able to split it into two phases. One phase that builds all the artifacts and then a 2nd phase that tries to create multiple tarballs in parallel.

**Comment by matthiaskrgr** [comment]
I wonder if we could model steps like artifact compression as a build.rs of a dummy crate so that we could let cargo figure out the scheduling.

miri dist would have a build.rs which depends on "miri" the built binary, in the build.rs, we would then execute the compression of the tar.xz, the actual crate which the build.rs depends on would just be an empty main.

Something like ninja that is able to build all sorts of parts of the huge dep tree in parallel, but for rust crates

**Comment by the8472** [comment]
I think directly using `jobserver` to distribute/wait for cpu resources among multiple running compression jobs is better. cargo doesn't help because it doesn't make those tokens transparent to the build script.
